### PR TITLE
Changes for zAuction  to allow multiple ERC20 tokens

### PR DESCRIPTION
**Pull request by JamesEarle**
- Allow a plethora of ways to check allowance and set approval
- Allow setting / getting of the payment token used in a domain network (e.g. 0://wilder)
- Modify configuration types, use zAuction SDK for all domains not per instance

Working on updating tests.

### CODE REVIEW

File: src/configuration/index.ts [hunk 1]
```diff
@@ -16,18 +16,9 @@ export const mainnetConfiguration = (
...
-        config: zAuctionConfiguration(provider, "mainnet") as zAuction.Config,
-      } as zAuctionRoute,
-      {
-        uriPattern: ".+",
-        // Use default values
-        config: zAuctionConfiguration(provider, "mainnet") as zAuction.Config,
-      } as zAuctionRoute,
-    ],
+    zAuction: {
+      ...zAuctionConfiguration(provider, "mainnet")
```
Comment (Remscar) [inline src/configuration/index.ts:20]: Shouldn't we pass in the hub here?

Comment (JamesEarle) [reply src/configuration/index.ts:20]: Modified this

File: package.json [hunk 2]
```diff
@@ -54,7 +55,6 @@
     "lolex": "6.0.0"
   },
   "peerDependencies": {
-    "@zero-tech/zauction-sdk": ">=0.0.55",
```
Comment (Remscar) [inline package.json:57]: Make sure to re-add this but `>=0.1.0` when the zAuction SDK PR goes in

File: src/configuration/index.ts [hunk 3]
```diff
@@ -12,22 +11,18 @@ const mainnetBasicController = "0xa05Ae774Da859943B7B859cd2A6aD9F5f1651d6a";
...
-        // Use default values
-        config: zAuctionConfiguration(provider, "mainnet") as zAuction.Config,
-      } as zAuctionRoute,
-      {
-        uriPattern: ".+",
-        // Use default values
-        config: zAuctionConfiguration(provider, "mainnet") as zAuction.Config,
-      } as zAuctionRoute,
-    ],
+    zAuction: {
```
Comment (Remscar) [inline src/configuration/index.ts:23]: nit: `zAuction: zAuctionConfiguration(mainnetConfig)`

File: tsconfig.build.json [hunk 4]
```diff
@@ -1,7 +1,6 @@
 {
   "compilerOptions": {
     "declaration": true,
-    "declarationMap": true,
```
Comment (Remscar) [inline tsconfig.build.json:4]: Does this need to be removed? It's nice to have this for debugging.

Comment (JamesEarle) [reply tsconfig.build.json:4]: Must have been deleted my something automatically? will make sure it's re added

File: test/e2e.test.ts [hunk 5]
```diff
@@ -1,16 +1,12 @@
...
 import { Registrar, ZNSHub } from "../src/contracts/types";
 import { Config, Instance } from "../src/types";
 import { rinkebyConfiguration } from "../src/configuration";
 import { createInstance } from "../src";
 import { expect } from "chai";
-import { buildExecutionContext } from "graphql/execution/execute";
-import { Bid } from "@zero-tech/zauction-sdk";
-import { assert } from "console";
-
+// import { ethers } from  "hardhat";
```
Comment (RyanCross) [inline test/e2e.test.ts]: remove?

File: package.json [hunk 6]
```diff
@@ -14,13 +14,14 @@
     "build": "tsc --build --verbose ./tsconfig.build.json",
     "lint": "eslint --fix src",
     "style": "prettier --write ./src/**/*.ts",
-    "test": "mocha -r ts-node/register 'test/**/*.test.ts' --exit",
+    "test": "mocha -r ts-node/register 'test/**/*.test.ts' --exit --timeout 25000",
```
Comment (RyanCross) [inline package.json:17]: nit: it might be better to place this in a `.mocharc`

Comment (JamesEarle) [reply package.json]: good call, made this change

File: src/configuration/index.ts [hunk 7]
```diff
@@ -12,22 +11,18 @@ const mainnetBasicController = "0xa05Ae774Da859943B7B859cd2A6aD9F5f1651d6a";
 export const mainnetConfiguration = (
   provider: ethers.providers.Provider
 ): Config => {
+  const mainnetConfig: zAuctionConfig = {
+    web3Provider: provider,
+    network: "rinkeby",
```
Comment (Remscar) [inline src/configuration/index.ts:16]: wrong network?

**Review by Will-Hyde — APPROVED** [review]

File: src/index.ts [hunk 8]
```diff
@@ -226,278 +209,244 @@ export const createInstance = (config: Config): Instance => {
       return tx;
     },
     zauction: {
-      needsToApproveZAuctionToSpendTokens: async (
+      getPaymentTokenInfo: async (
+        paymentTokenAddress: string,
+        chainName: string
```
Comment (Remscar) [inline src/index.ts:214]: Could you infer the chain name based on how the instance was configured? So it's not a requirement to pass it in here?

File: src/types.ts [hunk 9]
```diff
@@ -32,8 +33,20 @@ export interface ZAuctionInstances {
   [registrarAddress: string]: zAuction.Instance;
 }
 
-export interface RouteUriToInstance {
-  [key: string]: ZAuctionInstances;
+export interface TokenAddressMapping {
+  [address: string]: Mapping;
```
Comment (Remscar) [inline src/types.ts:37]: This should be `Maybe<Mapping>`

where `Maybe<T> = T | undefined`

`Maybe` should already be defined somewhere though

File: src/types.ts [hunk 10]
```diff
@@ -32,8 +33,20 @@ export interface ZAuctionInstances {
...
+}
+
+export interface Mapping {
```
Comment (Remscar) [inline src/types.ts:40]: What is this a `Mapping` of? This is too generic of a name. It has a specific usage and the name should be more specific to that usage.

File: src/types.ts [hunk 11]
```diff
@@ -32,8 +33,20 @@ export interface ZAuctionInstances {
...
+  [address: string]: {
+    id: string,
```
Comment (Remscar) [inline src/types.ts:42]: Make this a separate interface instead of inline. This should also be `Maybe<T>`

File: src/types.ts [hunk 12]
```diff
@@ -32,8 +33,20 @@ export interface ZAuctionInstances {
...
+    name: string
+  };
+}
+
+export interface TokenInfo {
```
Comment (Remscar) [inline src/types.ts:47]: Nit: too generic of a name. What information is this? What's it's specific usage?
Seems like this is pricing information so maybe `TokenPricingInfo`

File: src/types.ts [hunk 13]
```diff
@@ -32,8 +33,20 @@ export interface ZAuctionInstances {
   [registrarAddress: string]: zAuction.Instance;
 }
 
-export interface RouteUriToInstance {
-  [key: string]: ZAuctionInstances;
+export interface TokenAddressMapping {
```
Comment (Remscar) [inline src/types.ts:36]: Nit: Too generic of a name. Make it more specific to the name

File: src/index.ts [hunk 14]
```diff
@@ -226,278 +209,244 @@ export const createInstance = (config: Config): Instance => {
...
+      ): Promise<TokenInfo> => {
+        const tokenInfo =
```
Comment (Remscar) [inline src/index.ts:216]: Nit: Refactor this logic into the `actions` folder

File: src/index.ts [hunk 15]
```diff
@@ -226,278 +209,244 @@ export const createInstance = (config: Config): Instance => {
...
+          price: tokenPriceUsd,
+          name: tokenInfo.name
+        } as TokenInfo
+      },
+      setPaymentTokenForDomain: async (
+        networkId: string,
+        paymentTokenAddress: string,
+        signer: ethers.Signer
+      ): Promise<ethers.ContractTransaction> => {
+        const hub = await getHubContract(config.provider, config.hub);
```
Comment (Remscar) [inline src/index.ts:236]: Nit: Refactor this logic into the `actions` folder

File: src/types.ts [hunk 16]
```diff
@@ -224,30 +231,101 @@ export interface Instance {
...
+     * specific token on their behalf using a given paymentTokenAddress
+     * @param paymentTokenAddress The address of an ERC-20 payment token
+     * @param account The user account that will use funds
+     * @param amount The amount to check how much the user has allowed
+     */
+    needsToApproveZAuctionToSpendTokensByPaymentToken(
+      paymentTokenAddress: string,
+      account: string,
+      amount: string
+    ): Promise<any>;
```
Comment (Remscar) [inline src/types.ts:297]: No `any`

File: src/actions/getPaymentTokenInfo.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  if(!tokenData) {
+    throw Error("Can't find information about that token")
+  }
+
+  const tokenPriceUsd = tokenData.data.market_data.current_price.usd;
+
+  return {
+    price: tokenPriceUsd,
+    name: tokenInfo.name
+  } as TokenPriceInfo
```
Comment (Remscar) [inline src/actions/getPaymentTokenInfo.ts:35]: Dangerous `as` cast

Comment (Remscar) [reply src/actions/getPaymentTokenInfo.ts:35]: `const priceInfo: TokenPriceInfo = { ... }` is the safer way

File: src/configuration/index.ts (hunk 7 above)
Comment (Remscar) [inline src/configuration/index.ts:16]: Still looks like the wrong network? This is mainnet config, why are we using rinkeby network?

File: src/index.ts [hunk 18]
```diff
@@ -226,278 +209,228 @@ export const createInstance = (config: Config): Instance => {
       return tx;
     },
     zauction: {
-      needsToApproveZAuctionToSpendTokens: async (
+      getPaymentTokenInfo: async (
```
Comment (Remscar) [inline src/index.ts:212]: If you want you can shorten this to:
```
getPaymentTokenInfo: actions.getPaymentTokenInfo,
```

File: src/types.ts [hunk 19]
```diff
@@ -32,8 +34,22 @@ export interface ZAuctionInstances {
   [registrarAddress: string]: zAuction.Instance;
 }
 
-export interface RouteUriToInstance {
-  [key: string]: ZAuctionInstances;
+export interface NetworksToAddresses {
```
Comment (Remscar) [inline src/types.ts:37]: This interface is called `NetworksToAddresses` but in the definition you are mapping `address` to `AddressesToTokenApiInfo`. I'm assuming you meant `[network: string]` ?

Also what is this mapping really of? What are these addresses? Think of `address` as a raw type like `number` or `string`. I don't think `NetworksToNumber` would be intuitive to a user.

File: src/types.ts [hunk 20]
```diff
@@ -32,8 +34,22 @@ export interface ZAuctionInstances {
...
+  [address: string]: Maybe<AddressesToTokenApiInfo>;
+}
+
+export interface AddressesToTokenApiInfo {
```
Comment (Remscar) [inline src/types.ts:41]: Nit: What are these addresses? As a developer here I know it's a token address, but from the interface it's not clear.
Also could we be more verbose on what `TokenApiInfo` is? This is info specifically for coin gecko so maybe we could call this `CoinGeckoApiParams`?

File: src/types.ts [hunk 21]
```diff
@@ -32,8 +34,22 @@ export interface ZAuctionInstances {
...
+  [address: string]: Maybe<TokenApiInfo>
+}
+
+export interface TokenApiInfo {
```
Comment (Remscar) [inline src/types.ts:45]: Could we move `NetworksToAddresses` `AddressesToTokenApiInfo` and `TokenApiInfo` to the `src/actions/getPaymentTokenInfo.ts` file? These seem very specific to a single operation that we do, and are only ever used internally. We should avoid putting internal/private types into our `types.ts` file (or module) since the user will be importing all of these and we don't want to confuse them with internal/implementation details which they have no control over.

File: src/utilities/domains.ts [hunk 22]
```diff
@@ -40,3 +41,29 @@ export const getAbsoluteDomainPath = (
   const fixedPath = `${rootDomain}.${domain}`;
   return fixedPath;
 };
+
+// ID must be the given ID from CoinGecko API.
+// You can get this value from a token's page under "API ID"
+// https://www.coingecko.com/en/coins/zero-tech
+export const tokenAddressToFriendlyName: NetworksToAddresses = {
```
Comment (Remscar) [inline src/utilities/domains.ts:48]: Could we move this into `src/actions/getPaymentTokenInfo.ts` since it's the only place we use it?

**Review by Remscar — APPROVED** [review]
Looks good after the comments I left are resolved
